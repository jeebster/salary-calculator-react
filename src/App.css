form {
  margin: 50px 0 50px 0;
}

label {
  margin: 0 20px 0 0;
}

.app {
  text-align: center;
}

.form-control:not(:last-child) {
  margin: 0 0 25px 0
}