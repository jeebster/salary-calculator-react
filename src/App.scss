@import "../node_modules/bulma/sass/utilities/_all.sass";
@import "../node_modules/bulma/sass/base/_all.sass";
@import "../node_modules/bulma/sass/elements/button.sass";
@import "../node_modules/bulma/sass/elements/container.sass";
@import "../node_modules/bulma/sass/elements/title.sass";
@import "../node_modules/bulma/sass/form/_all.sass";
@import "../node_modules/bulma/sass/components/navbar.sass";
@import "../node_modules/bulma/sass/layout/footer.sass";
@import "../node_modules/bulma/sass/layout/section.sass";

.app {
  text-align: center;
}

.calculated-total-text {
  color: #5fbb6c;
}