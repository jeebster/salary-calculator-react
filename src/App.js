import React, { Component } from 'react';
import './App.scss';

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      annualSalaryAmount: 0.0,
      calculatedTotal: 0,
      healthCareAmount: 0.0,
      investmentsAmount: 0.0,
      investmentsMatchPercentage: 0.0,
      taxPercentage: 0.0,
      miscBenefitsAmount: 0.0,
      workHoursPerWeek: 40,
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(event) {
    event.preventDefault();
    this.setState({
      [event.target.name] : parseFloat(event.target.value)
    });
  }

  handleSubmit(event) {
    event.preventDefault();

    const {
      annualSalaryAmount, healthCareAmount,
      investmentsAmount, investmentsMatchPercentage,
      taxPercentage, workHoursPerWeek, miscBenefitsAmount
    } = this.state;

    // base calculations
    const workHoursPerYear = workHoursPerWeek * 52 // weeks-per-year
    const investmentsMatchAmount = (investmentsAmount * (investmentsMatchPercentage / 100));
    const taxAdjustedSalaryAmount = annualSalaryAmount + (annualSalaryAmount * (taxPercentage / 100));
    const adjustedHealthcareAmount = healthCareAmount * 12; // benefits cost per-year
    const adjustedMiscBenefitsAmount = miscBenefitsAmount * 12; // benefits cost per-year
    const adjustedInvestmentsAmount = (investmentsAmount * 12) + (investmentsMatchAmount * 12) // total per-year

    const adjustedCompensationAmount =
      taxAdjustedSalaryAmount +
      adjustedHealthcareAmount +
      adjustedInvestmentsAmount +
      adjustedMiscBenefitsAmount;

    // ultimate calculation
    const calculatedTotal = adjustedCompensationAmount / workHoursPerYear;

    this.setState({
      calculatedTotal: calculatedTotal
    });
  }

  renderFormInput(label, inputName, inputValue) {
    return (
      <div className="field is-horizontal">
        <div className="field-label is-normal">
          <label className="label">{label}</label>
        </div>
        <div className="field-body">
          <div className="field">
            <p className="control">
              <input
                name={inputName}
                className="input"
                type="number"
                value={inputValue}
                onChange={this.handleChange}
                required
              />
            </p>
          </div>
        </div>
      </div>
    )
  }

  render() {
    const {
      annualSalaryAmount, calculatedTotal,
      healthCareAmount, investmentsAmount,
      investmentsMatchPercentage, workHoursPerWeek,
      taxPercentage, miscBenefitsAmount
    } = this.state;

    return (
      <div className="app">
        <header className="section">
          <div className="container">
            <h1 className="title">Compensation Rate Calculator</h1>
            <p className="subtitle">Calculate your equivalent annual salary to an hourly rate</p>
          </div>
        </header>

        <section className="section">
          <div className="container">
            <form onSubmit={this.handleSubmit}>
              {this.renderFormInput("Work Hours Per Week", "workHoursPerWeek", workHoursPerWeek)}
              {this.renderFormInput("Annual Salary Amount", "annualSalaryAmount", annualSalaryAmount)}
              {this.renderFormInput("Tax Rate", "taxPercentage", taxPercentage)}
              {this.renderFormInput("Monthly Healthcare Premium", "healthCareAmount", healthCareAmount)}
              {this.renderFormInput("Monthly 401k Investment Amount", "investmentsAmount", investmentsAmount)}
              {this.renderFormInput("Employer 401k Investment Match", "investmentsMatchPercentage", investmentsMatchPercentage)}
              {this.renderFormInput("Monthly Miscellaneous Benefits Amount", "miscBenefitsAmount", miscBenefitsAmount)}

              <div className="field is-horizontal">
                <div className="field-label">
                  {/* Left empty for spacing */}
                </div>
                <div className="field-body">
                  <div className="field">
                    <div className="control">
                      <input
                        className="button is-primary"
                        type="submit"
                        value="Calculate"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </form>
          </div>

          <section className="section">
            <div className="container">
              <h2 className="title is-4">
                Your Rate: <span className="calculated-total-text">{`$${calculatedTotal.toFixed(2)}`}</span>
              </h2>
            </div>
          </section>
        </section>

        <footer className="footer">
          by <a href="https://gregjmorrison.com">Greg Morrison</a>
        </footer>
      </div>
    )
  }
}

export default App;
