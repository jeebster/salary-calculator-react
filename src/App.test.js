import React from 'react';
import { cleanup, fireEvent, render } from '@testing-library/react';
import App from './App';

afterEach(() => {
  cleanup();
})

test('renders correctly', () => {
  const app = render(<App />);
  expect(app).toMatchSnapshot();
});

test('it calculates correctly', () => {
  // calculation setup
  const annualSalaryAmount = 120000;
  const healthCareAmount = 550;
  const investmentsAmount = 1500;
  const investmentsMatchPercentage = 4;
  const taxPercentage = 25;
  const miscBenefitsAmount = 150;
  const workHoursPerWeek = 40;

  // base calculations
  const workHoursPerYear = workHoursPerWeek * 52; // weeks-per-year
  const investmentsMatchAmount = (investmentsAmount * (investmentsMatchPercentage / 100));
  const taxAdjustedSalaryAmount = annualSalaryAmount + (annualSalaryAmount * (taxPercentage / 100));
  const adjustedHealthcareAmount = healthCareAmount * 12; // benefits cost per-year
  const adjustedMiscBenefitsAmount = miscBenefitsAmount * 12; // benefits cost per-year
  const adjustedInvestmentsAmount = (investmentsAmount * 12) + (investmentsMatchAmount * 12); // total per-year

  const adjustedCompensationAmount =
    taxAdjustedSalaryAmount +
    adjustedHealthcareAmount +
    adjustedInvestmentsAmount +
    adjustedMiscBenefitsAmount;

  // ultimate calculation
  const expectedTotal = `$${(adjustedCompensationAmount / workHoursPerYear).toFixed(2)}`;

  // render
  const { getByLabelText } = render(<App />);

  // update inputs
  const inputMap = {
    workHoursPerWeek : workHoursPerWeek,
    annualSalaryAmount : annualSalaryAmount,
    healthCareAmount : healthCareAmount,
    investmentsAmount : investmentsAmount,
    investmentsMatchPercentage : investmentsMatchPercentage,
    taxPercentage : taxPercentage,
    miscBenefitsAmount : miscBenefitsAmount,
  };
  Object.keys(inputMap).forEach(key => {
    const node = getByLabelText(key);
    fireEvent.change(node, {
      target: {
        value: inputMap[key]
      }
    });
  });

  // submit form
  const submitNode = getByLabelText('submit');
  fireEvent.click(submitNode);

  // assertions
  const resultNode = getByLabelText('calculated-total');
  expect(resultNode.value).toBe(expectedTotal);
})