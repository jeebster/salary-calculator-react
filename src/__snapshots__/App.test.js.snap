// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="app"
      >
        <header
          class="section"
        >
          <div
            class="container"
          >
            <h1
              class="title"
            >
              Compensation Rate Calculator
            </h1>
            <p
              class="subtitle"
            >
              Calculate your equivalent annual salary to an hourly rate
            </p>
          </div>
        </header>
        <section
          class="section"
        >
          <div
            class="container"
          >
            <form>
              <div
                class="field is-horizontal"
              >
                <div
                  class="field-label is-normal"
                >
                  <label
                    class="label"
                  >
                    Work Hours Per Week
                  </label>
                </div>
                <div
                  class="field-body"
                >
                  <div
                    class="field"
                  >
                    <p
                      class="control"
                    >
                      <input
                        aria-label="workHoursPerWeek"
                        class="input"
                        name="workHoursPerWeek"
                        required=""
                        type="number"
                        value="40"
                      />
                    </p>
                  </div>
                </div>
              </div>
              <div
                class="field is-horizontal"
              >
                <div
                  class="field-label is-normal"
                >
                  <label
                    class="label"
                  >
                    Annual Salary Amount
                  </label>
                </div>
                <div
                  class="field-body"
                >
                  <div
                    class="field"
                  >
                    <p
                      class="control"
                    >
                      <input
                        aria-label="annualSalaryAmount"
                        class="input"
                        name="annualSalaryAmount"
                        required=""
                        type="number"
                        value="0"
                      />
                    </p>
                  </div>
                </div>
              </div>
              <div
                class="field is-horizontal"
              >
                <div
                  class="field-label is-normal"
                >
                  <label
                    class="label"
                  >
                    Tax Rate
                  </label>
                </div>
                <div
                  class="field-body"
                >
                  <div
                    class="field"
                  >
                    <p
                      class="control"
                    >
                      <input
                        aria-label="taxPercentage"
                        class="input"
                        name="taxPercentage"
                        required=""
                        type="number"
                        value="0"
                      />
                    </p>
                  </div>
                </div>
              </div>
              <div
                class="field is-horizontal"
              >
                <div
                  class="field-label is-normal"
                >
                  <label
                    class="label"
                  >
                    Monthly Healthcare Premium
                  </label>
                </div>
                <div
                  class="field-body"
                >
                  <div
                    class="field"
                  >
                    <p
                      class="control"
                    >
                      <input
                        aria-label="healthCareAmount"
                        class="input"
                        name="healthCareAmount"
                        required=""
                        type="number"
                        value="0"
                      />
                    </p>
                  </div>
                </div>
              </div>
              <div
                class="field is-horizontal"
              >
                <div
                  class="field-label is-normal"
                >
                  <label
                    class="label"
                  >
                    Monthly 401k Investment Amount
                  </label>
                </div>
                <div
                  class="field-body"
                >
                  <div
                    class="field"
                  >
                    <p
                      class="control"
                    >
                      <input
                        aria-label="investmentsAmount"
                        class="input"
                        name="investmentsAmount"
                        required=""
                        type="number"
                        value="0"
                      />
                    </p>
                  </div>
                </div>
              </div>
              <div
                class="field is-horizontal"
              >
                <div
                  class="field-label is-normal"
                >
                  <label
                    class="label"
                  >
                    Employer 401k Investment Match
                  </label>
                </div>
                <div
                  class="field-body"
                >
                  <div
                    class="field"
                  >
                    <p
                      class="control"
                    >
                      <input
                        aria-label="investmentsMatchPercentage"
                        class="input"
                        name="investmentsMatchPercentage"
                        required=""
                        type="number"
                        value="0"
                      />
                    </p>
                  </div>
                </div>
              </div>
              <div
                class="field is-horizontal"
              >
                <div
                  class="field-label is-normal"
                >
                  <label
                    class="label"
                  >
                    Monthly Miscellaneous Benefits Amount
                  </label>
                </div>
                <div
                  class="field-body"
                >
                  <div
                    class="field"
                  >
                    <p
                      class="control"
                    >
                      <input
                        aria-label="miscBenefitsAmount"
                        class="input"
                        name="miscBenefitsAmount"
                        required=""
                        type="number"
                        value="0"
                      />
                    </p>
                  </div>
                </div>
              </div>
              <div
                class="field is-horizontal"
              >
                <div
                  class="field-label"
                />
                <div
                  class="field-body"
                >
                  <div
                    class="field"
                  >
                    <div
                      class="control"
                    >
                      <input
                        aria-label="submit"
                        class="button is-primary"
                        type="submit"
                        value="Calculate"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </form>
          </div>
          <section
            class="section"
          >
            <div
              class="container"
            >
              <h2
                class="title is-4"
              >
                Your Rate: 
                <span
                  aria-label="calculated-total"
                  class="calculated-total-text"
                >
                  $0.00
                </span>
              </h2>
            </div>
          </section>
        </section>
        <footer
          class="footer"
        >
          by 
          <a
            href="https://gregjmorrison.com"
          >
            Greg Morrison
          </a>
        </footer>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="app"
    >
      <header
        class="section"
      >
        <div
          class="container"
        >
          <h1
            class="title"
          >
            Compensation Rate Calculator
          </h1>
          <p
            class="subtitle"
          >
            Calculate your equivalent annual salary to an hourly rate
          </p>
        </div>
      </header>
      <section
        class="section"
      >
        <div
          class="container"
        >
          <form>
            <div
              class="field is-horizontal"
            >
              <div
                class="field-label is-normal"
              >
                <label
                  class="label"
                >
                  Work Hours Per Week
                </label>
              </div>
              <div
                class="field-body"
              >
                <div
                  class="field"
                >
                  <p
                    class="control"
                  >
                    <input
                      aria-label="workHoursPerWeek"
                      class="input"
                      name="workHoursPerWeek"
                      required=""
                      type="number"
                      value="40"
                    />
                  </p>
                </div>
              </div>
            </div>
            <div
              class="field is-horizontal"
            >
              <div
                class="field-label is-normal"
              >
                <label
                  class="label"
                >
                  Annual Salary Amount
                </label>
              </div>
              <div
                class="field-body"
              >
                <div
                  class="field"
                >
                  <p
                    class="control"
                  >
                    <input
                      aria-label="annualSalaryAmount"
                      class="input"
                      name="annualSalaryAmount"
                      required=""
                      type="number"
                      value="0"
                    />
                  </p>
                </div>
              </div>
            </div>
            <div
              class="field is-horizontal"
            >
              <div
                class="field-label is-normal"
              >
                <label
                  class="label"
                >
                  Tax Rate
                </label>
              </div>
              <div
                class="field-body"
              >
                <div
                  class="field"
                >
                  <p
                    class="control"
                  >
                    <input
                      aria-label="taxPercentage"
                      class="input"
                      name="taxPercentage"
                      required=""
                      type="number"
                      value="0"
                    />
                  </p>
                </div>
              </div>
            </div>
            <div
              class="field is-horizontal"
            >
              <div
                class="field-label is-normal"
              >
                <label
                  class="label"
                >
                  Monthly Healthcare Premium
                </label>
              </div>
              <div
                class="field-body"
              >
                <div
                  class="field"
                >
                  <p
                    class="control"
                  >
                    <input
                      aria-label="healthCareAmount"
                      class="input"
                      name="healthCareAmount"
                      required=""
                      type="number"
                      value="0"
                    />
                  </p>
                </div>
              </div>
            </div>
            <div
              class="field is-horizontal"
            >
              <div
                class="field-label is-normal"
              >
                <label
                  class="label"
                >
                  Monthly 401k Investment Amount
                </label>
              </div>
              <div
                class="field-body"
              >
                <div
                  class="field"
                >
                  <p
                    class="control"
                  >
                    <input
                      aria-label="investmentsAmount"
                      class="input"
                      name="investmentsAmount"
                      required=""
                      type="number"
                      value="0"
                    />
                  </p>
                </div>
              </div>
            </div>
            <div
              class="field is-horizontal"
            >
              <div
                class="field-label is-normal"
              >
                <label
                  class="label"
                >
                  Employer 401k Investment Match
                </label>
              </div>
              <div
                class="field-body"
              >
                <div
                  class="field"
                >
                  <p
                    class="control"
                  >
                    <input
                      aria-label="investmentsMatchPercentage"
                      class="input"
                      name="investmentsMatchPercentage"
                      required=""
                      type="number"
                      value="0"
                    />
                  </p>
                </div>
              </div>
            </div>
            <div
              class="field is-horizontal"
            >
              <div
                class="field-label is-normal"
              >
                <label
                  class="label"
                >
                  Monthly Miscellaneous Benefits Amount
                </label>
              </div>
              <div
                class="field-body"
              >
                <div
                  class="field"
                >
                  <p
                    class="control"
                  >
                    <input
                      aria-label="miscBenefitsAmount"
                      class="input"
                      name="miscBenefitsAmount"
                      required=""
                      type="number"
                      value="0"
                    />
                  </p>
                </div>
              </div>
            </div>
            <div
              class="field is-horizontal"
            >
              <div
                class="field-label"
              />
              <div
                class="field-body"
              >
                <div
                  class="field"
                >
                  <div
                    class="control"
                  >
                    <input
                      aria-label="submit"
                      class="button is-primary"
                      type="submit"
                      value="Calculate"
                    />
                  </div>
                </div>
              </div>
            </div>
          </form>
        </div>
        <section
          class="section"
        >
          <div
            class="container"
          >
            <h2
              class="title is-4"
            >
              Your Rate: 
              <span
                aria-label="calculated-total"
                class="calculated-total-text"
              >
                $0.00
              </span>
            </h2>
          </div>
        </section>
      </section>
      <footer
        class="footer"
      >
        by 
        <a
          href="https://gregjmorrison.com"
        >
          Greg Morrison
        </a>
      </footer>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
